six/thirteen.py

```diff
@@ -0,0 +1,66 @@
+from collections import deque
+
+
+def strings_are_same(str1, str2):
+    for c1, c2 in zip(str1, str2):
+        if c1 != c2:
+            return False
+    return True
+
+
+class RollingHasher:
+    """Following https://en.wikipedia.org/wiki/Rabin%E2%80%93Karp_algorithm"""
+    
+    def __init__(self, substring_len, base=256, prime_modulus=101):
+        self.substring_len = substring_len
+        self.base = base
+        self.prime_modulus = prime_modulus
+        self._char_q = deque()
+        self._hash = 0
+        
+    def _add_to_hash(self, c):
+        self._hash = (self._hash * self.base + ord(c)) % self.prime_modulus
+        self._char_q.append(c)
+        
+    def _remove_first_in_from_hash(self):
+        left_base_offset = 1
+        for i in range(self.substring_len - 1):
```

Comment:
What is the runtime complexity of your implementation of R-K?

Reply:
looks like O(n*k) time (due to having to compute the base offset with a for loop to avoid int overflow) and O(k) space; k = substring length, n = main string length

Reply:
Agreed. Can you think of a way to get O(k+n)?

Reply:
can't find anything online for rolling hash that avoids an x^k exponentiation. if i don't use the safeguard against int overflow and use an optimized exp function that part could be reduced to O(log k), so total of O(n log k). doesn't seem like other solutions consider the cost of exponentiation in the big O time complexity

Reply:
You're computing `left_base_offset` O(k) for each of the n letters in the main string == O(n*k).

But do you have to compute it so many times?